src/NServiceBus.Core/ObjectBuilder/ContainerComponent.cs

```diff
@@ -0,0 +1,124 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using Container;
+    using ObjectBuilder;
+    using ObjectBuilder.Common;
+    using Settings;
+
+    class ContainerComponent
+    {
+        public ContainerComponent(SettingsHolder settings)
+        {
+            this.settings = settings;
+        }
+
+        public IConfigureComponents ContainerConfiguration { get; private set; }
+
+        public IBuilder Builder { get; private set; }
+
+        public void UseContainer<T>(Action<ContainerCustomizations> customizations = null) where T : ContainerDefinition, new()
+        {
+            customizations?.Invoke(new ContainerCustomizations(settings));
+
+            UseContainer(typeof(T));
+        }
+
+        public void UseContainer(Type definitionType)
+        {
+            UseContainer(definitionType.Construct<ContainerDefinition>().CreateContainer(settings));
+        }
+
+        public void UseContainer(IContainer container)
+        {
+            customContainer = container;
+        }
+
+        public void UseExternallyManagedBuilder(IBuilder builder)
+        {
+            ownsContainer = false;
+            Builder = builder;
+        }
+
+        public void AddUserRegistration(Action<IConfigureComponents> registration)
+        {
+            userRegistrations.Add(registration);
+        }
+
+        public void InitializeWithExternallyManagedContainer(IConfigureComponents configureComponents)
```

Comment:
While this component is internal so we control the usage, I'm wondering if we need to be more defensive? For example do we need to protect against calling both `InitializeWithExternallyManagedContainer` and `InitializeWithInternallyManagedContainer`?

Reply:
I thought about that but couldn't find a way to get into that invalid state via the public API. I didn't feel that it was important enough to add a unit test?

Reply:
If there's no way to get into that state in the public API, then the danger here would be if some future code change introduced a way, and then we'd have nothing to protect against it.

If you don't think it's important enough to worry about, then I'm fine with it. Just wanted to call out that in isolation, there's nothing here that prevents a bad state.